cassiopeia/tools/branch_length_estimator.py

```diff
@@ -0,0 +1,325 @@
+    tree.
+    """
+
+    @abc.abstractmethod
+    def estimate_branch_lengths(self, tree: Tree) -> None:
+        r"""
+        Estimates the branch lengths of the tree.
+
+        Annotates the tree's nodes with their estimated age, and
+        the tree's branches with their estiamted lengths. Operates on the tree
+        in-place.
+
+        Args:
+            tree: The tree for which to estimate branch lengths.
+        """
+
+
+class IIDExponentialBLE(BranchLengthEstimator):
+    r"""
+    A simple branch length estimator that assumes that the characters evolve IID
+    over the phylogeny with the same cutting rate.
+
+    This estimator requires that the ancestral states are provided.
+
+    The optimization problem is a special kind of convex program called an
+    exponential cone program:
+    https://docs.mosek.com/modeling-cookbook/expo.html
+    Because it is a convex program, it can be readily solved.
+
+    Args:
+        minimum_branch_length: Estimated branch lengths will be constrained to
+            have at least this lenght.
+        l2_regularization: Consecutive branches will be regularized to have
+            similar length via an L2 penalty whose weight is given by
+            l2_regularization.
+        verbose: Verbosity level.
+
+    Attributes:
+        log_likelihood: The log-likelihood of the training data under the
+            estimated model.
+        log_loss: The log-loss of the training data under the estimated model.
+            This is the log likhelihood plus the regularization terms.
+    """
+
+    def __init__(
+        self,
+        minimum_branch_length: float = 0,
+        l2_regularization: float = 0,
+        verbose: bool = False,
+    ):
+        self.minimum_branch_length = minimum_branch_length
+        self.l2_regularization = l2_regularization
+        self.verbose = verbose
+
+    def estimate_branch_lengths(self, tree: Tree) -> None:
+        r"""
+        See base class. The only caveat is that this method raises if it fails
+        to solve the underlying optimization problem for any reason.
+
+        Raises:
+            cp.error.SolverError
+        """
+        # Extract parameters
+        minimum_branch_length = self.minimum_branch_length
+        l2_regularization = self.l2_regularization
+        verbose = self.verbose
+
+        # # Wrap the networkx DiGraph for goodies.
+        # tree = Tree(tree)
+
+        # # # # # Create variables of the optimization problem # # # # #
+        r_X_t_variables = dict(
+            [
+                (node_id, cp.Variable(name=f"r_X_t_{node_id}"))
+                for node_id in tree.nodes()
+            ]
+        )
+        time_increases_constraints = [
+            r_X_t_variables[parent]
+            >= r_X_t_variables[child] + minimum_branch_length
+            for (parent, child) in tree.edges()
+        ]
+        leaves_have_age_0_constraints = [
+            r_X_t_variables[leaf] == 0 for leaf in tree.leaves()
```

Comment:
Yes, this is the ultrametric condition. We don't have an estimator without this condition. As of today I don't think we need it, but if we do I guess we can just add this option to this estimator as an argument?

Reply:
I think it's totally fine to keep this condition in there. It might be worthwhile to add this condition, somehow, to the name of the estimator like `SimpleUltrametricBLE` ( I know this is super long, but does that idea make sense?). That way, the estimator can be extended later into another class that does _not_ use the Ultrametric condition.